### PR TITLE
Make Credit Roles required in Mint Song Flow

### PR DESCRIPTION
**Pull request by dmkirshon**
Add: 
- Yup validation check and display error for Creditors which have no Role selected during update or next Minting step submissions

Refactor: 
- Update imports and props into an alphabetic order to enhance readability

Note: There are quite a few refactor changes so I will comment on the Credit roles specific changes. If the alphabetic refactor doesn't make sense then I will roll back the changes and only include the Creditor Roles fix.

https://projectnewm.atlassian.net/browse/STUD-10

### CODE REVIEW

File: apps/studio/src/pages/home/library/EditSong.tsx [hunk 1]
```diff
@@ -42,13 +41,14 @@ const EditSong: FunctionComponent = () => {
   const { songId } = useParams<"songId">() as SongRouteParams;
 
   const { data: genres = [] } = useGetGenresQuery();
+  const { data: roles = [] } = useGetRolesQuery();
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/EditSong.tsx:44]: Credit Role Change

File: apps/studio/src/pages/home/library/EditSong.tsx [hunk 2]
```diff
@@ -250,22 +251,23 @@ const EditSong: FunctionComponent = () => {
...
 
   const validations = {
+    barcodeNumber: commonYupValidation.barcodeNumber,
+    barcodeType: commonYupValidation.barcodeType,
+    consentsToContract: commonYupValidation.consentsToContract,
+    copyrightOwner: commonYupValidation.copyright,
+    copyrightYear: commonYupValidation.year,
     coverArtUrl: commonYupValidation.coverArtUrl,
-    title: commonYupValidation.title,
+    creditors: commonYupValidation.creditors(roles),
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/EditSong.tsx:260]: Credit Role Change

File: apps/studio/src/pages/home/library/EditSong.tsx [hunk 3]
```diff
@@ -333,11 +335,12 @@ const EditSong: FunctionComponent = () => {
                 handleSubmit("basic-details", values, helpers),
               validationSchema: Yup.object().shape({
                 coverArtUrl: validations.coverArtUrl,
-                title: validations.title,
+                creditors: validations.creditors,
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/EditSong.tsx:334]: Credit Role Change

File: apps/studio/src/pages/home/library/MintSong.tsx [hunk 4]
```diff
@@ -154,16 +161,19 @@ const MintSong = () => {
...
-    owners: initialOwners,
+    consentsToContract: false,
     creditors: initialCreditors,
     featured: initialFeatured,
-    consentsToContract: false,
+    isMinting,
+    owners: initialOwners,
   };
 
+  const { data: roles = [] } = useGetRolesQuery();
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/MintSong.tsx:171]: Credit Role Change

File: apps/studio/src/pages/home/library/MintSong.tsx [hunk 5]
```diff
@@ -154,16 +161,19 @@ const MintSong = () => {
...
+
   const validationSchema = Yup.object().shape({
-    owners: commonYupValidation.owners,
     consentsToContract: Yup.bool().required("This field is required"),
+    creditors: commonYupValidation.creditors(roles),
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/MintSong.tsx:175]: Credit Role Change

File: apps/studio/src/pages/home/library/MintSong.tsx [hunk 6]
```diff
@@ -53,7 +60,7 @@ const MintSong = () => {
   const { wallet } = useConnectWallet();
   const { songId } = useParams<"songId">() as SongRouteParams;
 
-  const ownersRef = useRef<HTMLDivElement>(null);
+  const coCreatorsRef = useRef<HTMLDivElement>(null);
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/MintSong.tsx:63]: Credit Role Change

File: apps/studio/src/pages/home/library/MintSong.tsx [hunk 7]
```diff
@@ -284,7 +294,8 @@ const MintSong = () => {
           };
 
           scrollToError(errors, isSubmitting, [
-            { error: errors.owners, element: ownersRef.current },
+            { error: errors.creditors, element: coCreatorsRef.current },
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/MintSong.tsx:297]: Credit Role Change

File: apps/studio/src/pages/home/library/MintSong.tsx [hunk 8]
```diff
@@ -284,7 +294,8 @@ const MintSong = () => {
...
+            { error: errors.owners, element: coCreatorsRef.current },
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/MintSong.tsx]: Credit Role Change

File: apps/studio/src/pages/home/library/MintSong.tsx [hunk 9]
```diff
@@ -311,33 +322,41 @@ const MintSong = () => {
...
                           onChangeFeatured={handleChangeFeatured}
+                          onChangeOwners={handleChangeOwners}
+                          owners={values.owners}
                         />
                       )}
                     </Box>
 
                     {!!touched.owners && !!errors.owners && (
-                      <Box mt={0.5} ref={ownersRef}>
+                      <Box mt={0.5} ref={coCreatorsRef}>
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/MintSong.tsx]: Credit Role Change

File: apps/studio/src/pages/home/library/MintSong.tsx [hunk 10]
```diff
@@ -311,33 +322,41 @@ const MintSong = () => {
...
                     )}
 
+                    {!!touched.creditors && !!errors.creditors && (
+                      <Box mt={0.5} ref={coCreatorsRef}>
+                        <ErrorMessage>
+                          {errors.creditors as string}
+                        </ErrorMessage>
+                      </Box>
+                    )}
+
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/MintSong.tsx]: Credit Role Change

File: apps/studio/src/pages/home/uploadSong/BasicSongDetails.tsx [hunk 11]
```diff
@@ -55,10 +61,10 @@ const BasicSongDetails: FunctionComponent<BasicDonDetailsProps> = ({
   const { wallet } = useConnectWallet();
 
   const audioRef = useRef<HTMLDivElement>(null);
+  const coCreatorsRef = useRef<HTMLDivElement>(null);
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/uploadSong/BasicSongDetails.tsx:64]: Credit Role Change

File: apps/studio/src/pages/home/uploadSong/BasicSongDetails.tsx [hunk 12]
```diff
@@ -130,7 +136,8 @@ const BasicSongDetails: FunctionComponent<BasicDonDetailsProps> = ({
         error: errors.description,
         element: descriptionRef.current,
       },
-      { error: errors.owners, element: ownersRef.current },
+      { error: errors.creditors, element: coCreatorsRef.current },
+      { error: errors.owners, element: coCreatorsRef.current },
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/uploadSong/BasicSongDetails.tsx]: Credit Role Change

File: apps/studio/src/pages/home/uploadSong/BasicSongDetails.tsx [hunk 13]
```diff
@@ -274,35 +281,35 @@ const BasicSongDetails: FunctionComponent<BasicDonDetailsProps> = ({
           <Stack spacing={3}>
             <Stack spacing={1.5}>
               <Box
-                ref={ownersRef}
+                ref={coCreatorsRef}
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/uploadSong/BasicSongDetails.tsx]: Credit Role Change

File: apps/studio/src/pages/home/uploadSong/BasicSongDetails.tsx [hunk 14]
```diff
@@ -312,6 +319,12 @@ const BasicSongDetails: FunctionComponent<BasicDonDetailsProps> = ({
                   <ErrorMessage>{errors.owners as string}</ErrorMessage>
                 </Box>
               )}
+
+              {!!touched.creditors && !!errors.creditors && (
+                <Box mt={0.5}>
+                  <ErrorMessage>{errors.creditors as string}</ErrorMessage>
+                </Box>
+              )}
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/uploadSong/BasicSongDetails.tsx]: Credit Role Change

File: apps/studio/src/pages/home/uploadSong/UploadSong.tsx [hunk 15]
```diff
@@ -1,14 +1,15 @@
...
 import * as Yup from "yup";
-import { Box, Container } from "@mui/material";
-import { FormikHelpers, FormikValues } from "formik";
 import { commonYupValidation } from "../../../common";
-import { useExtractProperty } from "@newm-web/utils";
-import { Typography, WizardForm } from "@newm-web/elements";
 import {
   useGetGenresQuery,
   useGetLanguagesQuery,
+  useGetRolesQuery,
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/uploadSong/UploadSong.tsx:15]: Credit Role Change

File: apps/studio/src/pages/home/uploadSong/UploadSong.tsx [hunk 16]
```diff
@@ -146,31 +148,32 @@ const UploadSong: FunctionComponent = () => {
...
-    coverArtUrl: commonYupValidation.coverArtUrl,
     audio: commonYupValidation.audio,
-    title: commonYupValidation.title,
+    barcodeNumber: commonYupValidation.barcodeNumber,
+    barcodeType: commonYupValidation.barcodeType,
+    consentsToContract: commonYupValidation.consentsToContract,
+    copyrightOwner: commonYupValidation.copyright,
+    copyrightYear: commonYupValidation.year,
+    coverArtUrl: commonYupValidation.coverArtUrl,
+    creditors: commonYupValidation.creditors(roles),
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/uploadSong/UploadSong.tsx:158]: Credit Role Change

File: apps/studio/src/pages/home/uploadSong/UploadSong.tsx [hunk 17]
```diff
@@ -194,13 +197,14 @@ const UploadSong: FunctionComponent = () => {
               onSubmitStep: handleSongInfo,
               navigateOnSubmitStep: false,
               validationSchema: Yup.object().shape({
-                coverArtUrl: validations.coverArtUrl,
                 audio: validations.audio,
-                title: validations.title,
+                coverArtUrl: validations.coverArtUrl,
+                creditors: validations.creditors,
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/uploadSong/UploadSong.tsx:201]: Credit Role Change

File: apps/studio/src/pages/home/uploadSong/UploadSong.tsx [hunk 18]
```diff
@@ -18,23 +19,24 @@ import {
...
-import BasicSongDetails from "./BasicSongDetails";
 import AdvancedSongDetails from "./AdvancedSongDetails";
+import BasicSongDetails from "./BasicSongDetails";
+import ConfirmAgreement from "./ConfirmAgreement";
 
 const UploadSong: FunctionComponent = () => {
   const navigate = useNavigate();
 
   const { data: genres = [] } = useGetGenresQuery();
+  const { data: roles = [] } = useGetRolesQuery();
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/uploadSong/UploadSong.tsx:30]: Credit Role Change

File: apps/studio/src/common/formUtils.ts [hunk 19]
```diff
@@ -185,6 +185,15 @@ export const commonYupValidation = {
...
   publicationDate: Yup.date().max(new Date(), "Cannot be a future date"),
+  creditors: (roles: string[]) =>
+    Yup.array().when("isMinting", {
+      is: (value: boolean) => !!value,
+      then: Yup.array().test({
+        message: "Creditors must have a role",
+        test: (creditors = []) =>
+          creditors.every(({ role }) => roles.includes(role)),
+      }),
+    }),
```
Comment (dmkirshon) [inline apps/studio/src/common/formUtils.ts]: Credit Role Change

File: apps/studio/src/pages/home/library/EditSong.tsx [hunk 20]
```diff
@@ -1,14 +1,19 @@
...
 import { useNavigate, useParams } from "react-router";
-import { Button } from "@newm-web/elements";
-import { ProfileImage } from "@newm-web/elements";
-import { WizardForm } from "@newm-web/elements";
+import * as Yup from "yup";
+import { commonYupValidation } from "../../../common";
+import {
+  useGetGenresQuery,
+  useGetLanguagesQuery,
+  useGetRolesQuery,
```
Comment (dmkirshon) [inline apps/studio/src/pages/home/library/EditSong.tsx:17]: Credit Role Change

**Review by scandycuz — APPROVED** [review]

**Pull request merged**